common/flatpak-dir.c

```diff
@@ -4018,19 +4022,231 @@ repo_get_remote_collection_id (OstreeRepo *repo,
 
+static const SecretSchema *
+get_api_server_access_token_schema (void)
+{
+    static const SecretSchema schema = {
+        "com.endlessm.HackAccountToken", SECRET_SCHEMA_NONE,
+        {
+            {  "serial", SECRET_SCHEMA_ATTRIBUTE_STRING },
+            {  "NULL", 0 },
+        }
+    };
+    return &schema;
+}
+
+static char *
+get_api_server_access_token (GCancellable  *cancellable,
+                             GError       **error)
+{
+  g_autofree char *product_serial = NULL;
+  g_autoptr(GFile) product_serial_file = NULL;
+  g_autoptr(GMainContextPopDefault) context = NULL;
+  g_autoptr(GAsyncResult) lookup_result = NULL;
+
+  product_serial_file = g_file_new_for_path (PRODUCT_SERIAL_PATH);
+  if (!g_file_load_contents (product_serial_file,
+                             cancellable,
+                             &product_serial,
+                             NULL, NULL,
+                             error))
+    return NULL;
+
+  g_strchomp (product_serial);
+
+  context = flatpak_main_context_new_default ();
+
+  secret_password_lookup (get_api_server_access_token_schema (),
+                          cancellable, async_result_cb, &lookup_result,
+                          "serial", product_serial,
+                          NULL);
+
+  while (lookup_result == NULL)
+    g_main_context_iteration (context, TRUE);
+
+  return secret_password_lookup_finish (lookup_result, error);
+}
+
+static char *
+get_repo_bearer_token (FlatpakDir    *dir,
+                       const char    *remote_url,
+                       const char    *api_server_url,
+                       const char    *ref_to_fetch,
+                       GCancellable  *cancellable,
+                       GError       **error)
+{
+  g_autofree char *access_token = NULL;
+  g_autofree char *bearer_header = NULL;
+  g_autofree char *token_url = NULL;
+  g_autofree char *token_url_checksum = NULL;
+  g_autofree char *token_cache_filename = NULL;
+  char *token_buffer = NULL;
+  gsize token_size;
+  g_autoptr(GFile) cache_dir = NULL;
+  g_autoptr(GFile) token_cache_file = NULL;
+  g_autoptr(GFileInfo) token_cache_file_info = NULL;
+  GTimeVal token_cache_mtime;
+  GTimeVal current_time;
+  g_autoptr(GMainContext) context = NULL;
+  g_autoptr(SoupRequestHTTP) request = NULL;
+  g_autoptr(SoupMessage) http_message = NULL;
+  g_autoptr(GAsyncResult) request_send_result = NULL;
+  g_autoptr(GInputStream) request_stream = NULL;
+  g_autoptr(GAsyncResult) request_read_all_result = NULL;
+
+  if (api_server_url != NULL)
+    {
+      access_token = get_api_server_access_token (cancellable, error);
+      if (!access_token)
+        {
+          g_prefix_error (error, "Unable to get API server access token: ");
+          return NULL;
+        }
+
+      token_url = g_strconcat (api_server_url,
+                               "/token?ref=",
+                               ref_to_fetch,
+                               "&repo_url=",
+                               remote_url,
```

Comment:
These parameters should be URI-escaped before building a URI with them. `SoupURI` provides an API for building URIs.